File: clients/java/src/main/java/com/lauradb/client/Collection.java
package com.lauradb.client;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.CompletableFuture;

/**
 * Represents a collection in LauraDB.
 *
 * <p>Example usage:
 * <pre>{@code
 * Collection users = client.collection("users");
 *
 * // Insert a document
 * Map<String, Object> doc = new HashMap<>();
 * doc.put("name", "Alice");
 * doc.put("age", 30);
 * String id = users.insertOne(doc);
 *
 * // Find documents
 * List<Map<String, Object>> results = users.find(
 *     Query.builder().gte("age", 25).build()
 * );
 * }</pre>
 */
public class Collection {
    private final LauraDBClient client;
    private final String name;

    Collection(LauraDBClient client, String name) {
        this.client = client;
        this.name = name;
    }

    /**
     * Get the collection name.
     *
     * @return the collection name
     */
    public String getName() {
        return name;
    }

    /**
     * Insert a single document.
     *
     * @param document the document to insert
     * @return the inserted document's ID
     * @throws IOException if the request fails
     */
    public String insertOne(Map<String, Object> document) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("document", document);

        JsonObject response = client.request("POST", "/" + name + "/_doc", body);
        return response.get("id").getAsString();
    }

    /**
     * Insert multiple documents.
     *
     * @param documents the documents to insert
     * @return list of inserted document IDs
     * @throws IOException if the request fails
     */
    @SuppressWarnings("unchecked")
    public List<String> insertMany(List<Map<String, Object>> documents) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("documents", documents);

        JsonObject response = client.request("POST", "/" + name + "/_bulk", body);
        return client.getGson().fromJson(response.get("ids"), List.class);
    }

    /**
     * Find a single document by ID.
     *
     * @param id the document ID
     * @return the document, or null if not found
     * @throws IOException if the request fails
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> findById(String id) throws IOException {
        try {
            JsonObject response = client.request("GET", "/" + name + "/_doc/" + id, null);
            return client.getGson().fromJson(response.get("document"), Map.class);
        } catch (IOException e) {
            if (e.getMessage().contains("404") || e.getMessage().contains("not found")) {
                return null;
            }
            throw e;
        }
    }

    /**
     * Find documents matching a query.
     *
     * @param query the query filter
     * @return list of matching documents
     * @throws IOException if the request fails
     */
    public List<Map<String, Object>> find(Query query) throws IOException {
        return find(query, new FindOptions());
    }

    /**
     * Find documents with options.
     *
     * @param query the query filter
     * @param options find options (projection, sort, limit, skip)
     * @return list of matching documents
     * @throws IOException if the request fails
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> find(Query query, FindOptions options) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());

        if (options.getProjection() != null) {
            body.put("projection", options.getProjection());
        }
        if (options.getSort() != null) {
            body.put("sort", options.getSort());
        }
        if (options.getLimit() > 0) {
            body.put("limit", options.getLimit());
        }
        if (options.getSkip() > 0) {
            body.put("skip", options.getSkip());
        }

        JsonObject response = client.request("POST", "/" + name + "/_query", body);
        JsonArray documents = response.getAsJsonArray("documents");

        List<Map<String, Object>> results = new ArrayList<>();
        for (JsonElement doc : documents) {
            results.add(client.getGson().fromJson(doc, Map.class));
        }
        return results;
    }

    /**
     * Find all documents in the collection.
     *
     * @return list of all documents
     * @throws IOException if the request fails
     */
    public List<Map<String, Object>> findAll() throws IOException {
        return find(Query.empty());
    }

    /**
     * Find one document matching a query.
     *
     * @param query the query filter
     * @return the first matching document, or null if none found
     * @throws IOException if the request fails
     */
    public Map<String, Object> findOne(Query query) throws IOException {
        FindOptions options = new FindOptions().limit(1);
        List<Map<String, Object>> results = find(query, options);
        return results.isEmpty() ? null : results.get(0);
    }

    /**
     * Count documents matching a query.
     *
     * @param query the query filter
     * @return the count of matching documents
     * @throws IOException if the request fails
     */
    public long count(Query query) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());

        JsonObject response = client.request("POST", "/" + name + "/_count", body);
        return response.get("count").getAsLong();
    }

    /**
     * Update a single document.
     *
     * @param query the query filter to find the document
     * @param update the update operations
     * @return the number of modified documents
     * @throws IOException if the request fails
     */
    public long updateOne(Query query, Map<String, Object> update) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());
        body.put("update", update);

        JsonObject response = client.request("POST", "/" + name + "/_update", body);
        return response.get("modified").getAsLong();
    }

    /**
     * Update a document by ID.
     *
     * @param id the document ID
     * @param update the update operations
     * @return the number of modified documents
     * @throws IOException if the request fails
     */
    public long updateById(String id, Map<String, Object> update) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("update", update);

        JsonObject response = client.request("PUT", "/" + name + "/_doc/" + id, body);
        return response.get("modified").getAsLong();
    }

    /**
     * Update multiple documents.
     *
     * @param query the query filter
     * @param update the update operations
     * @return the number of modified documents
     * @throws IOException if the request fails
     */
    public long updateMany(Query query, Map<String, Object> update) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());
        body.put("update", update);
        body.put("multi", true);

        JsonObject response = client.request("POST", "/" + name + "/_update", body);
        return response.get("modified").getAsLong();
    }

    /**
     * Delete a single document.
     *
     * @param query the query filter
     * @return the number of deleted documents
     * @throws IOException if the request fails
     */
    public long deleteOne(Query query) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());

        JsonObject response = client.request("POST", "/" + name + "/_delete", body);
        return response.get("deleted").getAsLong();
    }

    /**
     * Delete a document by ID.
     *
     * @param id the document ID
     * @return the number of deleted documents
     * @throws IOException if the request fails
     */
    public long deleteById(String id) throws IOException {
        JsonObject response = client.request("DELETE", "/" + name + "/_doc/" + id, null);
        return response.get("deleted").getAsLong();
    }

    /**
     * Delete multiple documents.
     *
     * @param query the query filter
     * @return the number of deleted documents
     * @throws IOException if the request fails
     */
    public long deleteMany(Query query) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("filter", query.getFilter());
        body.put("multi", true);

        JsonObject response = client.request("POST", "/" + name + "/_delete", body);
        return response.get("deleted").getAsLong();
    }

    /**
     * Execute an aggregation pipeline.
     *
     * @param pipeline the aggregation pipeline
     * @return list of aggregation results
     * @throws IOException if the request fails
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> aggregate(Aggregation pipeline) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("pipeline", pipeline.getPipeline());

        JsonObject response = client.request("POST", "/" + name + "/_aggregate", body);
        JsonArray results = response.getAsJsonArray("results");

        List<Map<String, Object>> output = new ArrayList<>();
        for (JsonElement result : results) {
            output.add(client.getGson().fromJson(result, Map.class));
        }
        return output;
    }

    /**
     * Create an index on a field.
     *
     * @param field the field to index
     * @return an IndexBuilder for further configuration
     */
    public IndexBuilder createIndex(String field) {
        return new IndexBuilder(this, field);
    }

    /**
     * Create a compound index on multiple fields.
     *
     * @param fields the fields to index
     * @return an IndexBuilder for further configuration
     */
    public IndexBuilder createCompoundIndex(List<String> fields) {
        return new IndexBuilder(this, fields);
    }

    /**
     * List all indexes in the collection.
     *
     * @return list of index names
     * @throws IOException if the request fails
     */
    @SuppressWarnings("unchecked")
    public List<String> listIndexes() throws IOException {
        JsonObject response = client.request("GET", "/" + name + "/_index", null);
        return client.getGson().fromJson(response.get("indexes"), List.class);
    }

    /**
     * Drop an index.
     *
     * @param field the indexed field
     * @throws IOException if the request fails
     */
    public void dropIndex(String field) throws IOException {
        client.request("DELETE", "/" + name + "/_index/" + field, null);
    }

    /**
     * Asynchronously insert a document.
     *
     * @param document the document to insert
     * @return a CompletableFuture with the document ID
     */
    public CompletableFuture<String> insertOneAsync(Map<String, Object> document) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                return insertOne(document);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    /**
     * Asynchronously find documents.
     *
     * @param query the query filter
     * @return a CompletableFuture with the results
     */
    public CompletableFuture<List<Map<String, Object>>> findAsync(Query query) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                return find(query);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }
}

File: clients/python/lauradb/collection.py
"""
LauraDB Collection - Interface for collection operations
"""
from typing import Dict, Any, List, Optional, Union


class Collection:
    """
    Collection object for performing CRUD operations.

    Args:
        client: LauraDB client instance
        name: Collection name

    Example:
        >>> users = client.collection('users')
        >>> users.insert_one({'name': 'Alice', 'age': 30})
    """

    def __init__(self, client, name: str):
        self.client = client
        self.name = name
        self._base_path = f"/collections/{name}"

    def insert_one(self, document: Dict[str, Any]) -> str:
        """
        Insert a single document into the collection.

        Args:
            document: Document to insert

        Returns:
            Inserted document ID

        Example:
            >>> doc_id = users.insert_one({'name': 'Alice', 'age': 30})
            >>> print(doc_id)
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/insert",
            body={"document": document}
        )
        return response.get("result", {}).get("id", "")

    def insert_many(self, documents: List[Dict[str, Any]]) -> List[str]:
        """
        Insert multiple documents into the collection.

        Args:
            documents: List of documents to insert

        Returns:
            List of inserted document IDs

        Example:
            >>> ids = users.insert_many([
            ...     {'name': 'Alice', 'age': 30},
            ...     {'name': 'Bob', 'age': 25}
            ... ])
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/insert-many",
            body={"documents": documents}
        )
        return response.get("result", {}).get("ids", [])

    def find_one(
        self,
        filter: Optional[Dict[str, Any]] = None,
        projection: Optional[Dict[str, int]] = None
    ) -> Optional[Dict[str, Any]]:
        """
        Find a single document matching the filter.

        Args:
            filter: Query filter (default: {})
            projection: Field projection (1 = include, 0 = exclude)

        Returns:
            Matching document or None

        Example:
            >>> doc = users.find_one({'name': 'Alice'})
            >>> print(doc['age'])
        """
        body = {"filter": filter or {}}
        if projection:
            body["projection"] = projection

        response = self.client._request(
            "POST",
            f"{self._base_path}/find-one",
            body=body
        )
        return response.get("result", {}).get("document")

    def find(
        self,
        filter: Optional[Dict[str, Any]] = None,
        projection: Optional[Dict[str, int]] = None,
        sort: Optional[Dict[str, int]] = None,
        skip: Optional[int] = None,
        limit: Optional[int] = None
    ) -> List[Dict[str, Any]]:
        """
        Find multiple documents matching the filter.

        Args:
            filter: Query filter (default: {})
            projection: Field projection (1 = include, 0 = exclude)
            sort: Sort specification (1 = ascending, -1 = descending)
            skip: Number of documents to skip
            limit: Maximum number of documents to return

        Returns:
            List of matching documents

        Example:
            >>> docs = users.find(
            ...     {'age': {'$gte': 25}},
            ...     projection={'name': 1, 'age': 1},
            ...     sort={'age': -1},
            ...     limit=10
            ... )
        """
        body = {"filter": filter or {}}
        if projection:
            body["projection"] = projection
        if sort:
            body["sort"] = sort
        if skip is not None:
            body["skip"] = skip
        if limit is not None:
            body["limit"] = limit

        response = self.client._request(
            "POST",
            f"{self._base_path}/find",
            body=body
        )
        return response.get("result", {}).get("documents", [])

    def count(self, filter: Optional[Dict[str, Any]] = None) -> int:
        """
        Count documents matching the filter.

        Args:
            filter: Query filter (default: {})

        Returns:
            Number of matching documents

        Example:
            >>> count = users.count({'age': {'$gte': 25}})
            >>> print(f"Found {count} users")
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/count",
            body={"filter": filter or {}}
        )
        return response.get("result", {}).get("count", 0)

    def update_one(
        self,
        filter: Dict[str, Any],
        update: Dict[str, Any]
    ) -> bool:
        """
        Update a single document matching the filter.

        Args:
            filter: Query filter
            update: Update operations (must use update operators like $set)

        Returns:
            True if a document was updated

        Example:
            >>> users.update_one(
            ...     {'name': 'Alice'},
            ...     {'$set': {'age': 31}}
            ... )
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/update-one",
            body={"filter": filter, "update": update}
        )
        return response.get("result", {}).get("modified", 0) > 0

    def update_many(
        self,
        filter: Dict[str, Any],
        update: Dict[str, Any]
    ) -> int:
        """
        Update multiple documents matching the filter.

        Args:
            filter: Query filter
            update: Update operations (must use update operators like $set)

        Returns:
            Number of documents updated

        Example:
            >>> count = users.update_many(
            ...     {'age': {'$lt': 18}},
            ...     {'$set': {'minor': True}}
            ... )
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/update-many",
            body={"filter": filter, "update": update}
        )
        return response.get("result", {}).get("modified", 0)

    def delete_one(self, filter: Dict[str, Any]) -> bool:
        """
        Delete a single document matching the filter.

        Args:
            filter: Query filter

        Returns:
            True if a document was deleted

        Example:
            >>> users.delete_one({'name': 'Alice'})
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/delete-one",
            body={"filter": filter}
        )
        return response.get("result", {}).get("deleted", 0) > 0

    def delete_many(self, filter: Dict[str, Any]) -> int:
        """
        Delete multiple documents matching the filter.

        Args:
            filter: Query filter

        Returns:
            Number of documents deleted

        Example:
            >>> count = users.delete_many({'age': {'$lt': 18}})
            >>> print(f"Deleted {count} users")
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/delete-many",
            body={"filter": filter}
        )
        return response.get("result", {}).get("deleted", 0)

    def aggregate(self, pipeline: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Execute an aggregation pipeline.

        Args:
            pipeline: List of aggregation stages

        Returns:
            List of aggregation results

        Example:
            >>> results = users.aggregate([
            ...     {'$match': {'age': {'$gte': 18}}},
            ...     {'$group': {
            ...         '_id': '$city',
            ...         'avgAge': {'$avg': '$age'},
            ...         'count': {'$count': {}}
            ...     }},
            ...     {'$sort': {'avgAge': -1}}
            ... ])
        """
        response = self.client._request(
            "POST",
            f"{self._base_path}/aggregate",
            body={"pipeline": pipeline}
        )
        return response.get("result", {}).get("documents", [])

    def create_index(
        self,
        field: str,
        unique: bool = False,
        sparse: bool = False,
        name: Optional[str] = None
    ) -> bool:
        """
        Create a B+ tree index on a field.

        Args:
            field: Field name to index
            unique: Whether the index should enforce uniqueness
            sparse: Whether to index only documents with the field
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> users.create_index('email', unique=True)
        """
        body = {
            "field": field,
            "unique": unique,
            "sparse": sparse
        }
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes",
            body=body
        )
        return response.get("ok", False)

    def create_compound_index(
        self,
        fields: List[str],
        unique: bool = False,
        name: Optional[str] = None
    ) -> bool:
        """
        Create a compound index on multiple fields.

        Args:
            fields: List of field names
            unique: Whether the index should enforce uniqueness
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> users.create_compound_index(['city', 'age'], name='city_age_idx')
        """
        body = {
            "fields": fields,
            "unique": unique
        }
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes/compound",
            body=body
        )
        return response.get("ok", False)

    def create_text_index(
        self,
        fields: List[str],
        name: Optional[str] = None
    ) -> bool:
        """
        Create a text index for full-text search.

        Args:
            fields: List of text field names
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> posts.create_text_index(['title', 'content'], name='posts_text')
        """
        body = {"fields": fields}
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes/text",
            body=body
        )
        return response.get("ok", False)

    def create_geo_index(
        self,
        field: str,
        geo_type: str = "2d",
        name: Optional[str] = None
    ) -> bool:
        """
        Create a geospatial index.

        Args:
            field: Field name containing coordinates
            geo_type: Index type ('2d' for planar, '2dsphere' for spherical)
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> locations.create_geo_index('coordinates', geo_type='2dsphere')
        """
        body = {
            "field": field,
            "geoType": geo_type
        }
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes/geo",
            body=body
        )
        return response.get("ok", False)

    def create_ttl_index(
        self,
        field: str,
        expire_after_seconds: int,
        name: Optional[str] = None
    ) -> bool:
        """
        Create a TTL (Time-To-Live) index for automatic document expiration.

        Args:
            field: Field name containing timestamp
            expire_after_seconds: Seconds after which documents expire
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> sessions.create_ttl_index('createdAt', expire_after_seconds=3600)
        """
        body = {
            "field": field,
            "expireAfterSeconds": expire_after_seconds
        }
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes/ttl",
            body=body
        )
        return response.get("ok", False)

    def create_partial_index(
        self,
        field: str,
        filter_expr: Dict[str, Any],
        unique: bool = False,
        name: Optional[str] = None
    ) -> bool:
        """
        Create a partial index that indexes only documents matching a filter.

        Args:
            field: Field name to index
            filter_expr: Filter expression for partial indexing
            unique: Whether the index should enforce uniqueness
            name: Optional index name

        Returns:
            True if successful

        Example:
            >>> users.create_partial_index(
            ...     'email',
            ...     filter_expr={'active': True},
            ...     unique=True
            ... )
        """
        body = {
            "field": field,
            "filterExpression": filter_expr,
            "unique": unique
        }
        if name:
            body["name"] = name

        response = self.client._request(
            "POST",
            f"{self._base_path}/indexes/partial",
            body=body
        )
        return response.get("ok", False)

    def list_indexes(self) -> List[Dict[str, Any]]:
        """
        List all indexes in the collection.

        Returns:
            List of index information

        Example:
            >>> indexes = users.list_indexes()
            >>> for idx in indexes:
            ...     print(idx['name'])
        """
        response = self.client._request(
            "GET",
            f"{self._base_path}/indexes"
        )
        return response.get("result", {}).get("indexes", [])

    def drop_index(self, name: str) -> bool:
        """
        Drop an index by name.

        Args:
            name: Index name

        Returns:
            True if successful

        Example:
            >>> users.drop_index('email_1')
        """
        response = self.client._request(
            "DELETE",
            f"{self._base_path}/indexes/{name}"
        )
        return response.get("ok", False)

    def stats(self) -> Dict[str, Any]:
        """
        Get collection statistics.

        Returns:
            Dictionary containing collection statistics

        Example:
            >>> stats = users.stats()
            >>> print(stats['documentCount'])
        """
        response = self.client._request(
            "GET",
            f"{self._base_path}/stats"
        )
        return response.get("result", {})

    def __repr__(self) -> str:
        """String representation of the collection."""
        return f"Collection(name='{self.name}')"
